OneSignalSDK/unittest/src/test/java/com/test/onesignal/DeviceTypeTestsRunner.java

```diff
@@ -0,0 +1,128 @@
+        TestHelpers.beforeTestSuite();
+        StaticResetHelper.saveStaticValues();
+    }
+
+    @Before
+    public void beforeEachTest() throws Exception {
+        TestHelpers.beforeTestInitAndCleanup();
+        OneSignal.setAppContext(RuntimeEnvironment.application);
+    }
+
+    @Test
+    public void noAvailablePushChannels_defaultsToAndroid() {
+        assertEquals(DEVICE_TYPE_ANDROID, getDeviceType());
+    }
+
+    @Test
+    public void onlyADM_isFireOS() {
+        ShadowOSUtils.supportsADM = true;
+        assertEquals(DEVICE_TYPE_FIREOS, getDeviceType());
+    }
+
+    @Test
+    public void onlyFCM_isAndroid() {
+        ShadowOSUtils.hasFCMLibrary = true;
+        assertEquals(DEVICE_TYPE_ANDROID, getDeviceType());
+    }
+
+    @Test
+    public void FCMAndGMSEnabled_isAndroid() {
+        ShadowOSUtils.isGMSInstalledAndEnabled = true;
+        ShadowOSUtils.hasFCMLibrary = true;
+        assertEquals(DEVICE_TYPE_ANDROID, getDeviceType());
+    }
+
+    @Test
+    public void onlyGCM_isAndroid() {
+        ShadowOSUtils.hasGCMLibrary = true;
+        assertEquals(DEVICE_TYPE_ANDROID, getDeviceType());
+    }
+
+    @Test
+    public void GCMAndGMSEnabled_isAndroid() {
+        ShadowOSUtils.isGMSInstalledAndEnabled = true;
+        ShadowOSUtils.hasGCMLibrary = true;
+        assertEquals(DEVICE_TYPE_ANDROID, getDeviceType());
+    }
+
+    @Test
+    public void supportedHMS_isHuawei() {
+        ShadowOSUtils.isHMSCoreInstalledAndEnabled = true;
+        ShadowOSUtils.hasHMSAvailability = true;
+        ShadowOSUtils.hasHMSPushKitLibrary = true;
+
+        assertEquals(DEVICE_TYPE_HUAWEI, getDeviceType());
+    }
+
+    @Test
+    public void supportsFCMAndHMS_PreferAndroid() {
+        ShadowOSUtils.isGMSInstalledAndEnabled = true;
+        ShadowOSUtils.hasFCMLibrary = true;
+
+        ShadowOSUtils.isHMSCoreInstalledAndEnabled = true;
+        ShadowOSUtils.hasHMSAvailability = true;
+        ShadowOSUtils.hasHMSPushKitLibrary = true;
+
+        // Prefer Google Services over Huawei if both available
+        assertEquals(DEVICE_TYPE_ANDROID, getDeviceType());
+    }
+
+    @Test
+    public void hasFCMButNoGMSOnDeviceAndHasHMS_isHuawei() {
+        ShadowOSUtils.isGMSInstalledAndEnabled = false;
+        ShadowOSUtils.hasFCMLibrary = true;
+
+        ShadowOSUtils.isHMSCoreInstalledAndEnabled = true;
+        ShadowOSUtils.hasHMSAvailability = true;
+        ShadowOSUtils.hasHMSPushKitLibrary = true;
+
+        // Use HMS since device does not have the "Google Play services" app or it is disabled
+        assertEquals(DEVICE_TYPE_HUAWEI, getDeviceType());
+    }
+
+    @Test
+    public void supportsFCMAndADM_PreferADM() {
+        ShadowOSUtils.isGMSInstalledAndEnabled = true;
+        ShadowOSUtils.hasFCMLibrary = true;
+
+        ShadowOSUtils.supportsADM = true;
+
+        // Prefer ADM as if available it will always be native to the device
+        assertEquals(DEVICE_TYPE_FIREOS, getDeviceType());
+    }
+}

```

Comment:
Just an overall comment for these `UnitTests`, maybe we should offer some steps explaining what each line truly means or create more specific test names.

This one is a good example and offers clarity
`supportsFCMAndADM_PreferADM`
but even could be `testDevice_withFCMAndADMSupport_prefersADM`

however, this one
`supportedHMS_isHuawei` 
could be
`testDevice_withHMSSupport_prefersHMS`

I'm probably being to critical though since it only took me a few seconds to figure out the pattern here. I guess this could stay the same since I recognized the consistent pattern, but wanted to bring it up to see what you thought.

Reply:
Adding `testDevice_` seems reducend if prefixing each test with it. 

The pattern I was going for was `state_expectedResult`.

I agree that we should narrow down a naming pattern, since the scope of this problem is limited to HMS only lets defer